### DISCUSSION TITLE
Install missing tools

### DISCUSSION THREAD

**Discussion by abitrolly** [evilmartians/lefthook#350] [Ideas]
I know it is probably not the goal of `lefthook`, but.. sometimes `lefthook` fails for new users, because they don't have tools installed.

https://gitlab.com/gitlab-org/gitlab-development-kit/-/issues/1584

Even if it is not the goal, discussing what is available today to cover this scenario would be beneficial. [`sg setup` is doing this](https://docs.sourcegraph.com/dev/background-information/sg/reference#sg-setup) - checking dependencies and installing them.

1. `asdf` - requires shell to be reloaded? PostgreSQL is installed from source, takes an hour.
2. `nix` - complicated. Not clear how to switch versions for testing.
3. `brew` - not clear how to pin versions.
4. `whalebrew` - looks promising, might be a way to go

The best solution is to provide tools in isolated environment, so `whalebrew` might be the best answer. Need to check that though. I don't like when SDKs are making intrusive changes to my system. These are sometimes hard to debug, so it is nice to limit all modifications only to the project `lefthook` is installed in.

**Comment by mrexox** [comment]
Hey! I got your point. I think the most natural way is to use `Gemfile` or `package.json` to handle dependencies (and `lefthook` binary also).  

But there is a feature lefthook supports for `lefthook-local.yml` - [`{cmd}`](https://github.com/evilmartians/lefthook/blob/master/docs/configuration.md#cmd-template) template. This allows to wrap the command from `lefthook.yml` into any other command, e.g., `docker run`. So, it might be a workaround - keep all dependencies in a special Docker environment and use it for running `lefthook` (and maybe other development tasks. See [**dip**](https://github.com/bibendi/dip) utility, which provides a tool for dockerized development environment.)

P.S.

There is an option called [`fail_text`](https://github.com/evilmartians/lefthook/blob/master/docs/configuration.md#fail_text) which can be used for the installation instructions, like `please make sure you have <tool> installed or install it with a "asdf install ..." command`. At least some information about dependencies might be helpful.

P.P.S.

And after all you can use [`scripts`](https://github.com/evilmartians/lefthook/blob/master/docs/configuration.md#script) which check that the required tools are installed or install the specific version.
